genesis/src/lib.rs

```diff
@@ -93,64 +112,65 @@ impl RawGenesisTransaction {
+    /// Prepare genesis block.
     ///
     /// # Errors
     /// If executor couldn't be read from provided path
-    pub fn build_and_sign(self, genesis_key_pair: &KeyPair) -> Result<GenesisBlock> {
+    pub fn build(self) -> Result<GenesisBlock> {
         let executor = get_executor(&self.executor.0)?;
-        let genesis = build_and_sign_genesis(
+        let genesis = build_genesis(
             self.instructions,
             executor,
             self.chain,
-            genesis_key_pair,
+            &self.genesis.signatory(),
             self.topology,
         );
         Ok(genesis)
     }
 }
 
-fn build_and_sign_genesis(
+fn build_genesis(
     instructions: Vec<InstructionBox>,
     executor: Executor,
     chain_id: ChainId,
-    genesis_key_pair: &KeyPair,
+    genesis_public_key: &PublicKey,
     topology: Vec<PeerId>,
 ) -> GenesisBlock {
-    let transactions = build_transactions(instructions, executor, chain_id, genesis_key_pair);
-    let block = SignedBlock::genesis(transactions, genesis_key_pair.private_key(), topology);
+    let transactions = build_transactions(instructions, executor, chain_id, genesis_public_key);
+    let block = SignedBlock::genesis(transactions, topology);
     GenesisBlock(block)
 }
 
 fn build_transactions(
     instructions: Vec<InstructionBox>,
     executor: Executor,
     chain_id: ChainId,
-    genesis_key_pair: &KeyPair,
+    genesis_public_key: &PublicKey,
 ) -> Vec<SignedTransaction> {
     let upgrade_isi = Upgrade::new(executor).into();
     let transaction_executor =
-        build_transaction(vec![upgrade_isi], chain_id.clone(), genesis_key_pair);
+        build_transaction(vec![upgrade_isi], chain_id.clone(), genesis_public_key);
     if instructions.is_empty() {
         vec![transaction_executor]
     } else {
-        let transaction_instructions = build_transaction(instructions, chain_id, genesis_key_pair);
+        let transaction_instructions =
+            build_transaction(instructions, chain_id, genesis_public_key);
         vec![transaction_executor, transaction_instructions]
     }
 }
 
 fn build_transaction(
     instructions: Vec<InstructionBox>,
     chain_id: ChainId,
-    genesis_key_pair: &KeyPair,
+    genesis_public_key: &PublicKey,
 ) -> SignedTransaction {
-    let genesis_account_id = AccountId::new(
-        GENESIS_DOMAIN_ID.clone(),
-        genesis_key_pair.public_key().clone(),
-    );
-    TransactionBuilder::new(chain_id, genesis_account_id)
+    let genesis_account = AccountId::new(GENESIS_DOMAIN_ID.clone(), genesis_public_key.clone());
+    TransactionBuilder::new(chain_id, genesis_account)
         .with_instructions(instructions)
-        .sign(genesis_key_pair.private_key())
+        // There is no need to sign genesis transactions since we check it by hash,
+        // but our data model requires SignedTransaction to be signed,
+        // so we sign it with random key pair
+        .sign(KeyPair::random().private_key())
```

Comment:
Just to bring attention for other reviewers. This is a bit contentious, but I think it's a good approach.